Fixing #8011: full check of typability of terms from ltac signature.

herbelin opened the pull request:
**Kind:**  bug fix

Fixes #8011 (and possibly other similar bug reports).

@zimmi48 comment at [#8011](https://github.com/coq/coq/issues/8011#issuecomment-403096166) convinced me that it was unreasonable to deliberatedly leave a known source of anomaly. This PR fixes the problem, to the price of some (possible) overhead. I did not follow closely the evolution of the benchmark infrastructure. If someone can start a bench on code using ltac intensively, that would be great.

Added:  fixes #3459, fixes #3460, fixes #5378, fixes #5504.

herbelin commented:
Note: this PR shall need a little of extra work (failure in test-suite), but, at least, it is now scheduled and I shall be able to work on it when I get a bit more free time.

herbelin commented:
Note: This PR would fix also #5378 and #5504 but not #5505, because `type_of` does not check the well-typing of the substitution of evars. (And that would probably be costly to do.)

herbelin commented:
Note that times are bad.
```
┌──────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬──────────────────┐
│                          │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │    mem faults    │
│                          │                         │                                       │                                       │                         │                  │
│             package_name │     NEW     OLD PDIFF   │            NEW            OLD PDIFF   │            NEW            OLD PDIFF   │     NEW     OLD PDIFF   │ NEW OLD  PDIFF   │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│      coq-formal-topology │   54.55   54.68 -0.24 % │   148944426401   149339343325 -0.26 % │   223504362461   223517814221 -0.01 % │  478960  479184 -0.05 % │   0   0   +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│               coq-sf-plf │   70.78   70.83 -0.07 % │   195519429413   195613667160 -0.05 % │   283195619638   283673453237 -0.17 % │  516544  515856 +0.13 % │  22   0   +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│   coq-mathcomp-ssreflect │   62.40   62.42 -0.03 % │   171135171209   171162529017 -0.02 % │   242393467566   242552607785 -0.07 % │  531932  531880 +0.01 % │   0   0   +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│    coq-mathcomp-solvable │  217.39  217.23 +0.07 % │   601173861853   599931920275 +0.21 % │   857945787850   858655501649 -0.08 % │  859484  825924 +4.06 % │   0   0   +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│   coq-mathcomp-character │  289.63  289.30 +0.11 % │   803316623137   802513922603 +0.10 % │  1169516644161  1171772937149 -0.19 % │ 1056452 1112068 -5.00 % │ 228   5 +4460.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│    coq-mathcomp-fingroup │   81.92   81.80 +0.15 % │   225438915909   225373275162 +0.03 % │   334159599324   334266661962 -0.03 % │  588960  586740 +0.38 % │   0   0   +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│                coq-sf-lf │   38.36   38.30 +0.16 % │   105055359336   105263390603 -0.20 % │   172259917521   172707439350 -0.26 % │  422928  422876 +0.01 % │ 189   0   +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│              coq-bignums │   96.42   96.20 +0.23 % │   266468412094   265627119058 +0.32 % │   379829831339   378977183645 +0.22 % │  517176  516532 +0.12 % │   0   0   +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│             coq-compcert │  892.02  889.27 +0.31 % │  2474002961938  2463660060309 +0.42 % │  3553243610628  3548298849924 +0.14 % │ 1355748 1349888 +0.43 % │  28  63 -55.56 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│                  coq-vst │ 1712.71 1706.96 +0.34 % │  4759302723460  4739013775105 +0.43 % │  6477047713637  6472204782162 +0.07 % │ 2225696 2226276 -0.03 % │   7  24 -70.83 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│       coq-mathcomp-field │  466.38  464.67 +0.37 % │  1294666235623  1290110743979 +0.35 % │  2119409979701  2118859922310 +0.03 % │  797104  803468 -0.79 % │   0   0   +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│            coq-fiat-core │  129.67  129.16 +0.39 % │   360544757936   358718990298 +0.51 % │   478313538531   476731854565 +0.33 % │  496092  496716 -0.13 % │  19   1 +1800.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│              coq-unimath │ 1321.62 1316.21 +0.41 % │  3660817965566  3646065287384 +0.40 % │  6228075118219  6228341961388 -0.00 % │ 1007744 1007696 +0.00 % │  74   7 +957.14 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│               coq-sf-vfa │   43.95   43.74 +0.48 % │   121011962123   121261561430 -0.21 % │   192686905806   192954032066 -0.14 % │  533196  533112 +0.02 % │  22   1 +2100.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│     coq-mathcomp-algebra │  201.07  200.09 +0.49 % │   555866222186   553871511846 +0.36 % │   780115787481   780407185060 -0.04 % │  648576  644548 +0.62 % │   0   0   +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│ coq-mathcomp-real-closed │  185.89  184.83 +0.57 % │   515045162870   512407250951 +0.51 % │   778394272382   778506770687 -0.01 % │  829596  831252 -0.20 % │   5   0   +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│                coq-flocq │  278.49  276.04 +0.89 % │   772820297769   765839529020 +0.91 % │   991397809688   985978047308 +0.55 % │ 1244632 1244708 -0.01 % │  41 347 -88.18 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│           coq-coquelicot │  101.58  100.49 +1.08 % │   279871215008   277931252680 +0.70 % │   373045157915   371518024242 +0.41 % │  715908  712660 +0.46 % │ 298  15 +1886.67 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│   coq-mathcomp-odd-order │ 1434.02 1418.38 +1.10 % │  3983691850323  3940363086916 +1.10 % │  6708453583468  6702380924741 +0.09 % │ 1401880 1353744 +3.56 % │   9   4 +125.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│         coq-math-classes │  246.92  242.58 +1.79 % │   676989258333   666169264590 +1.62 % │   878360157324   865238053229 +1.52 % │  531508  531700 -0.04 % │  69   2 +3350.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│                 coq-hott │  331.46  325.32 +1.89 % │   919962596307   905324529134 +1.62 % │  1382093263642  1367313496864 +1.08 % │  611240  611316 -0.01 % │  22   0   +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│                coq-color │  749.75  730.98 +2.57 % │  2074040928799  2019637128671 +2.69 % │  2485585483235  2433785372253 +2.13 % │ 1387624 1387636 -0.00 % │   6  11 -45.45 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│               coq-geocoq │ 2920.47 2814.91 +3.75 % │  8103694586206  7825447057482 +3.56 % │ 12927883482105 12570278776081 +2.84 % │ 1387760 1395812 -0.58 % │  37  50 -26.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│                 coq-corn │ 1672.36 1561.38 +7.11 % │  4628847743451  4320685447374 +7.13 % │  6827676973339  6394861186092 +6.77 % │  920244  883648 +4.14 % │  62   0   +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼──────────────────┤
│          coq-fiat-crypto │ 6448.00 5889.23 +9.49 % │ 17913052430650 16355988151474 +9.52 % │ 29803304313287 27677547593254 +7.68 % │ 3194700 3206736 -0.38 % │ 219 790 -72.28 % │
└──────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴──────────────────┘
```
Don't know how to avoid these bad results though. Tracing when a variable has been renamed or overwritten so that we can save the full typing when the context has not changed...

herbelin commented:
For the record: the `transparent_abstract` and `4781` failures in the test-suite are related to the unnamed proof in `abstract` which is not known from the typer, the third one, `Case22`is maybe due to a bug in the compilation of pattern-matching. To be continued...

SkySkimmer commented:
The times look like they'd make integration difficult. Maybe it would be enough to do a well-formedness check (ie all variables are bound) before the retyping? There would still be late type errors, and maybe even anomalies (reduction got ill-formed term or the like) but at least the anomalies would be harder to get.

andres-erbsen commented:
Is this the kind of thing it would make sense to have an option for? E.g. `Set Ltac Constr Usage mode` where `mode` can be `"safe"` (this PR), `"notypecheck"` (SkySkimmer's suggestion) or `"nocheck"` (current behavior). I think I would use it on "safe" by default (assuming that mode did not have novel bugs like primitive projections or universe polymorphism), and perhaps sometimes use `"nocheck" very rarely when I find this to be an issue (but there are other performance reasons to avoid building large constrs in Ltac, so I don't really see when I would). I am pretty sure I would never use `"notypecheck"` given the other two options. Whether or not the option could be implemented without additional issues (and at reasonable dev time cost) is beyond me.

herbelin commented:
I rebased and, following @andres-erbsen's suggestion, I added an option `Set Ltac Constr Full Typing` which is on by default and which only tests presence of variable names in the non-full mode.

As a short-term fix, I believe that this is acceptable but I'm not fully convinced that this is the right approach though. See e.g. a discussion at [#5378](https://github.com/coq/coq/issues/5378#issuecomment-581068474).

Note also that full Ltac typing reveals some issues that #9710 would solve (see test-suite `Case22.v` in the PR. So, I'm also advocating for not delaying #9710 excessively!

JasonGross commented:
If it is on by default, then this PR should get a bench to evaluate performance impact.  (Though maybe it should wait until we make fiat-crypto build with the latest version of bedrock2, because otherwise we can't bench fiat-crypto, cc @jadephilipoom @andres-erbsen )

JasonGross commented:
And apparently this change makes fiat-parsers OOM?

ppedrot commented:
I think this is too dangerous a fix to be integrated. Note that Ltac2 suffers from the same problem as it uses essentially the same infrastructure, but I was thinking about having a dedicated safer mechanism relying on typing caches to circumvent the issues. This would require careful design first though.

herbelin commented:
@ppedrot:
> too dangerous a fix 

You mean in terms of complexity/efficiency or in terms of being too "dynamic" in retyping (in the sense that a term, say `x`, can be retypable by chance in a context, say `x:nat`, which has nothing to do with the context in which `x` was originally typable).

> relying on typing caches

Do you mean binding typing contexts to some unique keys, and being able to "decide" equality or inclusion of contexts?

ppedrot commented:
Ltac2 doesn't fix the issues per se, but because it features a type system, it makes it easy to provide both contradicting requirements in the same language:
- on the one hand, a safe, slow API that performs type-checking on the fly
- on the other hand, a fast, low-level one that trusts the user
It's a lot of work though, since the problem is already implicitly present in the ML API: in general we have no idea what is expected from the arguments given to ML tactics...

ppedrot commented:
As argued, the runtime consequences of this fix are too heavy to be ignored. I still think we can do something clever relying on caches, but in this view this PR is not doing the right thing. For the time being statu quo seems to prevail, so I am closing it.
